ros2_reduct_agent/recorder.py

```diff
@@ -0,0 +1,376 @@
+                    msgdef_text=msg_def.encoded_message_definition,
+                )
+                state.schema_by_type[msg_type_str] = schema
+                self.logger.debug(
+                    f"[{topic_name}] Registered schema for message type '{msg_type_str}'"
+                )
+
+            state.schemas_by_topic[topic_name] = schema
+
+    def make_topic_callback(self, topic_name: str):
+        """Generate a callback that writes the message to any relevant pipeline."""
+
+        def _topic_callback(message):
+            publish_time = self.get_publish_time(message, topic_name)
+            self.process_message(topic_name, message, publish_time)
+
+        return _topic_callback
+
+    def get_publish_time(self, message: Any, topic_name: str) -> int:
+        """Extract publish time from message (header.stamp or top-level stamp) in nanoseconds."""
+        if hasattr(message, "header") and hasattr(message.header, "stamp"):
+            return Time.from_msg(message.header.stamp).nanoseconds
+        elif hasattr(message, "stamp"):
+            return Time.from_msg(message.stamp).nanoseconds
+
+        if topic_name not in self.warned_topics:
+            self.logger.warn(
+                f"Message on topic '{topic_name}' has no timestamp. Using current time."
+            )
+            self.warned_topics.add(topic_name)
+
+        return self.get_clock().now().nanoseconds
+
+    #
+    # Message Processing
+    #
+    def process_message(self, topic_name: str, message: Any, publish_time: int):
+        """Process message for all pipelines that include the topic."""
+        for pipeline_name, state in self.pipeline_states.items():
+            if topic_name not in state.topics:
+                continue
+
+            if state.first_timestamp is None:
+                state.first_timestamp = publish_time
+
+            self.logger.debug(
+                f"Writing message to pipeline '{pipeline_name}' [{topic_name}]"
+            )
+
+            schema = state.schemas_by_topic.get(topic_name)
+            if not schema:
+                self.logger.warn(
+                    f"[{pipeline_name}] No schema registered for topic '{topic_name}'"
+                )
+                continue
+
+            state.writer.write_message(
+                topic=topic_name,
+                schema=schema,
+                message=message,
+                publish_time=publish_time,
+            )
+            state.current_size = state.buffer.tell()
+            split_size = self.pipeline_configs[pipeline_name].split_max_size_bytes
+            if split_size and state.current_size >= split_size:
+                self.upload_pipeline(pipeline_name, state)
+
+    #
+    # Pipeline Management and Upload
+    #
+    def make_timer_callback(self, pipeline_name: str):
+        """Return a callback that uploads the current pipeline state."""
+
+        def _timer_callback():
+            state = self.pipeline_states[pipeline_name]
```

Comment:
it's fragile, you put a state in the dictionary before the call. If you change the order of calls , it will be broken and hard to find. Pass the state as an argument